### PR TITLE
[WIP] Add bucket storage

### PR DESCRIPTION
**Pull request by masudur-rahman**
- [x] Process `Avatar` from `Bucket` (Google, Azure and S3)
- [x] Process `Avatar` from local storage

To store `Avatar` on bucket the following code must be on `app.ini` file. If `app.ini` doesn't contain the following section, the avatar will be stored on local storage by default.
```
[storage]
BUCKET      = gs://<gitea-storage-demo>
BUCKET_URL  = https://storage.googleapis.com/<gitea-storage-demo>
```
The `Bucket` must be public.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,10 +23,9 @@ require (
 	github.com/cznic/b v0.0.0-20181122101859-a26611c4d92d // indirect
 	github.com/cznic/mathutil v0.0.0-20181122101859-297441e03548 // indirect
 	github.com/cznic/strutil v0.0.0-20181122101858-275e90344537 // indirect
-	github.com/denisenkom/go-mssqldb v0.0.0-20181014144952-4e0d7dc8888f
-	github.com/dgrijalva/jwt-go v0.0.0-20161101193935-9ed569b5d1ac
+	github.com/denisenkom/go-mssqldb v0.0.0-20190121005146-b04fd42d9952
```
Comment (techknowlogick) [inline go.mod]: some of these changes seem unrelated. Could you include only the minimal required changes?

File: models/storage.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	// Bucket storage not set
+	if !com.IsFile(u.CustomAvatarPath()) {
+		return "", errors.Errorf("file doesn't exist")
+	}
+
+	return setting.AppSubURL + "/avatars/" + u.Avatar, nil
+}
+
+// GetAvatarLinkFromBucket provides avatar link from cloud storage
+func (u *User) GetAvatarLinkFromBucket() (string, error) {
```
Comment (lafriks) [inline models/storage.go]: Why bucket functions need to be exported?

**Comment by tamalsaha** [comment]
@techknowlogick , we (@masudur-rahman and I) tried to see if we can just vendor the gocloud.dev@0.13.0 without making any other changes in go.mod file. But it seems impossible to do so. We have cleaned up this pr into 3 commits. The first commit makes the desired code changes. The 2nd commit runs `go get -u =patch gocloud.dev@v0.13.0` and the 3rd commit runs `go mod tidy;go mod vendor`. We are using Go 0.12.5 . Please advise on how to proceed.

**Comment by lunny** [comment]
Please resolve the conflicts
